activesupport/lib/active_support/core_ext/enumerable.rb

```diff
@@ -17,12 +19,45 @@ module Enumerable
   # The default sum of an empty list is zero. You can override this default:
   #
   #  [].sum(Payment.new(0)) { |i| i.amount } # => Payment.new(0)
-  def sum(identity = nil, &block)
-    if block_given?
-      map(&block).sum(identity)
-    else
-      sum = identity ? inject(identity, :+) : inject(:+)
-      sum || identity || 0
+  if Enumerable.instance_methods(false).include?(:sum) && !(%w[a].sum rescue false)
+    alias :orig_enum_sum :sum
+
+    def sum(identity = nil, &block) #:nodoc:
+      if identity
+        orig_enum_sum(identity, &block)
+      elsif identity = _identity_for_sum
+        orig_enum_sum(identity, &block)
+      else
+        if block_given?
+          map(&block).sum(identity)
+        else
+          sum = identity ? inject(identity, :+) : inject(:+)
+          sum || identity || 0
+        end
+      end
+    end
+
+    private def _identity_for_sum #:nodoc:
+      return 0 if first.is_a?(Numeric)
+      ActiveSupport::Deprecation.warn(<<-MSG.squish)
+          Support for calling #sum on a non-numeric enumerable without an identity argument,
+          is deprecated and will soon be removed.
```

Comment:
Do we actually agree with this? It's nice that upstream has partially adopted the API and optimized it, but we shouldn't lose sight of the fact we added it to make code look better... it's possible we still prefer our original, more generous, API. I certainly do.

OTOH, if we have in fact changed our mind about the API, this deprecation should apply on earlier Ruby versions too.